ipa-core/src/protocol/ipa_prf/boolean_ops/sigmoid.rs

```diff
@@ -158,18 +159,96 @@ where
     ]))
 }
 
+// Sigmoid(
+//     Sum(i = 1..N, neuron(i) in last layer activation times edge weight connecting that neuron to this)
+//   )
+//
+
+// for i in 0..M-1 // For going through all layers
+// for j in 0..N-1 // Current layer
+// for k in 0..N-1 // For previous layer
+// neuron(i*N + j) += neuron((i-1)*N + k) * edge_weight(neuron((i)*N + j), neuron((i-1)*N + k))
+
+// M' neurons wide and here M is M'/N, L layers tall
+pub async fn neural_network<C, S, const M: usize, const N: usize, const MTimesN: usize>(
+    ctx: C,
+    last_layer_neurons: &[BitDecomposed<AdditiveShare<Boolean, N>>; M],
+    edge_weights: &[BitDecomposed<AdditiveShare<Boolean, N>>; M],
+)  -> Result<BitDecomposed<AdditiveShare<Boolean, N>>, Error>
+where
+    C: Context,
+    S: NBitStep,
+    Boolean: FieldSimd<N>,
+    AdditiveShare<Boolean, N>: BooleanProtocols<C, N>,
+    Boolean: FieldSimd<M>,
+    AdditiveShare<Boolean, M>: BooleanProtocols<C, M>,
+{    
+    // use super::step::MultiplicationStep as Step;
+    // for each layer we get M*M vector of edge_weights
+    let mut mults = ctx.parallel_join(zip(edge_weights.iter(), last_layer_neurons).enumerate().map(|(i, (edge_weight, neuron))| {
+        let ctx = ctx.narrow(&TwoHundredFiftySixBitOpStep::Bit(i));       
+    async move {
+        integer_mul::<_, S, N>(
+            ctx,
+            RecordId::FIRST,
+            &edge_weight,
+            &neuron,
+        )
+        .await
+        }
+    })).await?;
+    
+    let mut num = 0;
+    while mults.len() > 1 {
+        // Add each of the mults amongst themselves
+        for (a, b) in mults.iter().tuples() {
+        let (add_result, _) = integer_add::<_, S, N>(
+            ctx.narrow(&TwoHundredFiftySixBitOpStep::Bit(M+num)),      
+            RecordId::from(num),
+            &a,
+            &b,
+        )
+        .await?;
+        mults.push(add_result);
+        num += 1;
+        }
+       
+    }
```

Comment:
Andy already has code that does this (log(n) depth steps, adding each time and thereby dividing the length of the list by 2). Use `pub async fn aggregate_values<'ctx, 'fut, C, OV, const B: usize>(`